File: libs/east/src/expr/blob.ts
/**
 * Copyright (c) 2025 Elara AI Pty Ltd
 * Dual-licensed under AGPL-3.0 and commercial license. See LICENSE for details.
 */
import type { AST } from "../ast.js";
import { get_location } from "../location.js";
import { ArrayType, IntegerType, StringType, type BlobType, type EastType, type StructType } from "../types.js";
import { valueOrExprToAstTyped } from "./ast.js";
import { AstSymbol, Expr, FactorySymbol, type ToExpr } from "./expr.js";
import type { IntegerExpr } from "./integer.js";
import type { StringExpr } from "./string.js";
import type { ExprType } from "./types.js";
import type { BooleanExpr } from "./boolean.js";
import { equal, notEqual } from "./block.js";
import { CsvParseConfigType, csvParseOptionsToValue, type CsvParseOptions } from "../serialization/csv.js";
import type { ArrayExpr } from "./array.js";

/**
 * Expression representing binary blob values and operations.
 *
 * BlobExpr provides methods for working with binary data including size queries,
 * byte access, text encoding/decoding (UTF-8, UTF-16), and BEAST binary format operations.
 * Blobs are immutable sequences of bytes.
 *
 * @example
 * ```ts
 * // Encoding and decoding text
 * const encodeText = East.function([StringType], BlobType, ($, text) => {
 *   $.return(text.encodeUtf8());
 * });
 *
 * const decodeText = East.function([BlobType], StringType, ($, blob) => {
 *   $.return(blob.decodeUtf8());
 * });
 *
 * // Working with blob bytes
 * const getByte = East.function([BlobType, IntegerType], IntegerType, ($, blob, offset) => {
 *   $.return(blob.getUint8(offset));
 * });
 * ```
 */
export class BlobExpr extends Expr<BlobType> {
  constructor(ast: AST, createExpr: ToExpr) {
    super(ast.type as BlobType, ast, createExpr);
  }
  
  /**
   * Returns the size of the blob in bytes.
   *
   * @returns An IntegerExpr representing the number of bytes
   *
   * @example
   * ```ts
   * const getSize = East.function([BlobType], IntegerType, ($, blob) => {
   *   $.return(blob.size());
   * });
   * const compiled = East.compile(getSize.toIR(), []);
   * const text = "hello";
   * const blob = new TextEncoder().encode(text);
   * compiled(blob);  // 5n (5 bytes for "hello")
   * ```
   */
  size(): IntegerExpr {
    return this[FactorySymbol]({
      ast_type: "Builtin",
      type: IntegerType,
      location: get_location(),
      builtin: "BlobSize",
      type_parameters: [],
      arguments: [this[AstSymbol]],
    }) as IntegerExpr;
  }
  
  /**
   * Gets the byte value at the specified offset.
   *
   * @param offset - The zero-based byte offset to read from
   * @returns An IntegerExpr representing the unsigned 8-bit integer (0-255) at that offset
   *
   * @throws East runtime error if the offset is out of bounds (< 0 or >= blob size)
   *
   * @example
   * ```ts
   * const getByte = East.function([BlobType, IntegerType], IntegerType, ($, blob, offset) => {
   *   $.return(blob.getUint8(offset));
   * });
   * const compiled = East.compile(getByte.toIR(), []);
   * const blob = new Uint8Array([72, 101, 108, 108, 111]); // "Hello"
   * compiled(blob, 0n);  // 72n (ASCII 'H')
   * compiled(blob, 1n);  // 101n (ASCII 'e')
   * // compiled(blob, 10n) would throw error (out of bounds)
   * ```
   */
  getUint8(offset: IntegerExpr | bigint): IntegerExpr {
    const offsetAst = valueOrExprToAstTyped(offset, IntegerType);
    
    return this[FactorySymbol]({
      ast_type: "Builtin",
      type: IntegerType,
      location: get_location(),
      builtin: "BlobGetUint8",
      type_parameters: [],
      arguments: [this[AstSymbol], offsetAst],
    }) as any;
  }

  /**
   * Decodes the blob as UTF-8 text, returning a string.
   *
   * @returns A StringExpr containing the decoded text
   *
   * @throws East runtime error if the blob contains invalid UTF-8 sequences
   *
   * @remarks UTF-8 is a superset of ASCII. No byte-order mark (BOM) is expected.
   *
   * @example
   * ```ts
   * const decode = East.function([BlobType], StringType, ($, blob) => {
   *   $.return(blob.decodeUtf8());
   * });
   * const compiled = East.compile(decode.toIR(), []);
   * const blob = new TextEncoder().encode("hello");
   * compiled(blob);  // "hello"
   *
   * const emojiBlob = new TextEncoder().encode("😀");
   * compiled(emojiBlob);  // "😀"
   * ```
   */
  decodeUtf8(): StringExpr {
    return this[FactorySymbol]({
      ast_type: "Builtin",
      type: StringType,
      location: get_location(),
      builtin: "BlobDecodeUtf8",
      type_parameters: [],
      arguments: [this[AstSymbol]],
    }) as StringExpr;
  }

  /**
   * Decodes the blob as UTF-16 text, returning a string.
   *
   * @returns A StringExpr containing the decoded text
   *
   * @throws East runtime error if the blob contains invalid UTF-16 sequences
   *
   * @remarks UTF-16 is common on Windows. If no byte-order mark (BOM) is present, assumes little-endian.
   *
   * @example
   * ```ts
   * const decode = East.function([BlobType], StringType, ($, blob) => {
   *   $.return(blob.decodeUtf16());
   * });
   * const compiled = East.compile(decode.toIR(), []);
   * // Assume blob contains UTF-16 LE encoded "hello" with BOM
   * const blob = new Uint8Array([0xFF, 0xFE, 0x68, 0x00, 0x65, 0x00, 0x6C, 0x00, 0x6C, 0x00, 0x6F, 0x00]);
   * compiled(blob);  // "hello"
   * ```
   */
  decodeUtf16(): StringExpr {
    return this[FactorySymbol]({
      ast_type: "Builtin",
      type: StringType,
      location: get_location(),
      builtin: "BlobDecodeUtf16",
      type_parameters: [],
      arguments: [this[AstSymbol]],
    }) as StringExpr;
  }

  /**
   * Decodes a value from East's binary format (BEAST).
   *
   * @param type - The expected East type of the decoded value
   * @param version - The BEAST format version ('v1' or 'v2', default 'v1')
   * @returns An expression of the specified type containing the decoded value
   *
   * @throws East runtime error if the blob does not contain a valid BEAST encoding of the specified type
   *
   * @example
   * ```ts
   * const decode = East.function([BlobType], ArrayType(IntegerType), ($, blob) => {
   *   $.return(blob.decodeBeast(ArrayType(IntegerType), 'v2'));
   * });
   * const compiled = East.compile(decode.toIR(), []);
   * // Assume blob contains BEAST v2 encoding of [1n, 2n, 3n]
   * const encodedBlob = // ... BEAST encoded data ... ;
   * compiled(encodedBlob);  // [1n, 2n, 3n]
   * ```
   */
  decodeBeast<T extends EastType>(type: T, version: 'v1' | 'v2' = 'v1'): ExprType<T> {
    if (version === 'v1') {
      return this[FactorySymbol]({
        ast_type: "Builtin",
        type,
        location: get_location(),
        builtin: "BlobDecodeBeast",
        type_parameters: [type],
        arguments: [this[AstSymbol]],
      }) as ExprType<T>;
    } else if (version === 'v2') {
      return this[FactorySymbol]({
        ast_type: "Builtin",
        type,
        location: get_location(),
        builtin: "BlobDecodeBeast2",
        type_parameters: [type],
        arguments: [this[AstSymbol]],
      }) as ExprType<T>;
    } else {
      throw new Error(`Unsupported Beast version: ${version}`);
    }
  }

  /**
   * Decodes the blob as CSV data into an array of structs.
   *
   * @param structType - The struct type for each row
   * @param options - CSV parsing options
   * @returns An ArrayExpr containing the decoded structs
   *
   * @throws East runtime error if the CSV is malformed or doesn't match the struct type
   *
   * @example
   * ```ts
   * const PersonType = StructType({ name: StringType, age: IntegerType });
   *
   * const parseCsv = East.function([BlobType], ArrayType(PersonType), ($, blob) => {
   *   $.return(blob.decodeCsv(PersonType, { delimiter: ',' }));
   * });
   * const compiled = East.compile(parseCsv.toIR(), []);
   * const csv = new TextEncoder().encode("name,age\nAlice,30\nBob,25");
   * compiled(csv);  // [{ name: "Alice", age: 30n }, { name: "Bob", age: 25n }]
   * ```
   */
  decodeCsv<T extends StructType>(structType: T, options?: CsvParseOptions): ArrayExpr<T> {
    // Convert options to East config value
    const configValue = csvParseOptionsToValue(options);
    const configAst = valueOrExprToAstTyped(configValue, CsvParseConfigType);

    return this[FactorySymbol]({
      ast_type: "Builtin",
      type: ArrayType(structType),
      location: get_location(),
      builtin: "BlobDecodeCsv",
      type_parameters: [structType, CsvParseConfigType],
      arguments: [this[AstSymbol], configAst],
    }) as ArrayExpr<T>;
  }

  /**
   * Checks if this blob equals another blob (byte-by-byte comparison).
   *
   * @param other - The blob to compare against
   * @returns A BooleanExpr that is true if the blobs are identical
   *
   * @example
   * ```ts
   * const isEqual = East.function([BlobType, BlobType], BooleanType, ($, a, b) => {
   *   $.return(a.equals(b));
   * });
   * const compiled = East.compile(isEqual.toIR(), []);
   * const blob1 = new Uint8Array([1, 2, 3]);
   * const blob2 = new Uint8Array([1, 2, 3]);
   * const blob3 = new Uint8Array([1, 2, 4]);
   * compiled(blob1, blob2);  // true
   * compiled(blob1, blob3);  // false
   * ```
   */
  equals(other: BlobExpr | Uint8Array): BooleanExpr {
    return equal(this, other);
  }

  /**
   * Checks if this blob does not equal another blob.
   *
   * @param other - The blob to compare against
   * @returns A BooleanExpr that is true if the blobs are different
   *
   * @example
   * ```ts
   * const isNotEqual = East.function([BlobType, BlobType], BooleanType, ($, a, b) => {
   *   $.return(a.notEquals(b));
   * });
   * const compiled = East.compile(isNotEqual.toIR(), []);
   * const blob1 = new Uint8Array([1, 2, 3]);
   * const blob2 = new Uint8Array([1, 2, 4]);
   * compiled(blob1, blob2);  // true
   * compiled(blob1, blob1);  // false
   * ```
   */
  notEquals(other: BlobExpr | Uint8Array): BooleanExpr {
    return notEqual(this, other);
  }

  // ============================================================================
  // Aliases for comparison operations
  // ============================================================================

  /** Alias for {@link equals} */
  eq = this.equals;
  /** Alias for {@link equals} */
  equal = this.equals;

  /** Alias for {@link notEquals} */
  ne = this.notEquals;
  /** Alias for {@link notEquals} */
  notEqual = this.notEquals;
}

File: libs/east-py/packages/east-py/east/expression/expr/blob.py
#
# Copyright (c) 2025 Elara AI Pty Ltd
# Licensed under the Business Source License 1.1. See LICENSE.md for details.
#
"""``BlobExpression`` — TS ``BlobExpr`` (``libs/east/src/expr/blob.ts``)."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from east.expression.errors import ExpressionError
from east.expression.expr.base import Expression
from east.expression.lift import _lift
from east.expression.nodes import _builtin
from east.types.types import ArrayType, EastType, IntegerType, StringType

if TYPE_CHECKING:
    from east.expression.expr.array import ArrayExpression
    from east.expression.expr.integer import IntegerExpression
    from east.expression.expr.string import StringExpression


class BlobExpression(Expression):
    """A Blob-typed expression (immutable bytes): size, byte access, text
    decoding, and the BEAST / CSV decoders."""

    __slots__ = ()
    _kind = "Blob"

    def size(self) -> IntegerExpression:
        """Traced BlobSize: the number of bytes."""
        return self._expr(_builtin("BlobSize", IntegerType, [], [self.ir]), IntegerType)

    def get_uint8(self, offset: Any) -> IntegerExpression:
        """Traced BlobGetUint8: the byte at ``offset`` (0-255; an East runtime
        error when out of bounds)."""
        i = self._typed("get_uint8", offset, IntegerType)
        return self._expr(_builtin("BlobGetUint8", IntegerType, [], [self.ir, i.ir]), IntegerType)

    def decode_utf8(self) -> StringExpression:
        """Traced BlobDecodeUtf8 (an East runtime error on invalid UTF-8)."""
        return self._expr(_builtin("BlobDecodeUtf8", StringType, [], [self.ir]), StringType)

    def decode_utf16(self) -> StringExpression:
        """Traced BlobDecodeUtf16 (little-endian without a BOM)."""
        return self._expr(_builtin("BlobDecodeUtf16", StringType, [], [self.ir]), StringType)

    def decode_beast(self, typ: EastType, version: str = "v1") -> Expression:
        """Traced BlobDecodeBeast / BlobDecodeBeast2: decode East's binary
        format as a value of ``typ``. ``'v2'`` names the beast2 FAMILY — the
        decoder dispatches on the blob's magic byte, so every released
        container version reads."""
        if not isinstance(typ, EastType):
            raise ExpressionError(".decode_beast() takes an East type")
        if version == "v1":
            builtin = "BlobDecodeBeast"
        elif version == "v2":
            builtin = "BlobDecodeBeast2"
        else:
            raise ExpressionError(f"Unsupported Beast version: {version!r} (expected 'v1' or 'v2')")
        return self._expr(_builtin(builtin, typ, [typ], [self.ir]), typ)

    def open_beast(self, typ: EastType) -> Expression:
        """Traced BlobOpenBeast2 (TS ``openBeast``): open an indexed beast2 v5
        collection blob as a FROZEN lazy paged value of ``typ`` — ``size``,
        keyed reads and ``b.for_`` loops answer from the segment index, and
        anything else hydrates the whole value once (#657). ``typ`` must be an
        Array, Set or Dict type. A blob that cannot page (an index-less blob
        such as ``East.Blob.encode_beast(v, 'v2')`` writes, a v4 container, an
        element shape carrying a Ref or a function) decodes whole, frozen; a
        v5 header of another type is an East runtime error."""
        if not isinstance(typ, EastType) or typ.type not in ("Array", "Set", "Dict"):
            raise ExpressionError(".open_beast() takes an Array, Set or Dict type")
        return self._expr(_builtin("BlobOpenBeast2", typ, [typ], [self.ir]), typ)

    def decode_csv(self, struct_type: EastType, config: Any = None, **options: Any) -> ArrayExpression:
        """Traced BlobDecodeCsv: the CSV bytes as an Array of ``struct_type``
        rows. ``config`` is a ``CsvParseConfigType`` value; the keyword
        ``options`` are ``east.serialization.csv.csv_parse_config``'s
        (``delimiter=…``, ``has_header=…``, ``null_strings=…``, …)."""
        from east.serialization.csv import CsvParseConfigType, csv_parse_config

        if not isinstance(struct_type, EastType) or struct_type.type != "Struct":
            raise ExpressionError(".decode_csv() takes the row StructType")
        if config is not None and options:
            raise ExpressionError(".decode_csv() takes a config value OR keyword options, not both")
        cfg = _lift(config if config is not None else csv_parse_config(**options),
                    hint=CsvParseConfigType)
        out = ArrayType(struct_type)
        return self._expr(
            _builtin("BlobDecodeCsv", out, [struct_type, CsvParseConfigType], [self.ir, cfg.ir]),
            out,
        )
